File: nvim/language-examples/go/math/math.go
package math

import "math"

// go to definition on `math.Add` or `math.Circle` in main.go jumps here
// hover on any exported symbol shows its doc comment

// Add returns the sum of two integers.
func Add(a, b int) int {
    return a + b
}

// Shape is implemented by any type with an Area method.
// go to definition on the Shape interface in main.go jumps here
type Shape interface {
    Area() float64
}

type Circle struct {
    Radius float64
}

// find references on `Area` shows all implementations and call sites
func (c Circle) Area() float64 {
    return math.Pi * c.Radius * c.Radius
}

type Rectangle struct {
    Width, Height float64
}

func (r Rectangle) Area() float64 {
    return r.Width * r.Height
}

File: nvim/language-examples/rust/src/math.rs
// go to definition on `math::add` or `math::Shape` in main.rs jumps here
// find references on `area` shows all call sites across the project

pub fn add(a: i32, b: i32) -> i32 {
    a + b
}

pub trait Shape {
    // go to definition on `area` call in main.rs jumps here
    // hover shows the trait definition and all implementors
    fn area(&self) -> f64;
}

pub struct Circle {
    pub radius: f64,
}

pub struct Rectangle {
    pub width: f64,
    pub height: f64,
}

impl Shape for Circle {
    fn area(&self) -> f64 {
        std::f64::consts::PI * self.radius * self.radius
    }
}

impl Shape for Rectangle {
    fn area(&self) -> f64 {
        self.width * self.height
    }
}
